R/dryRun.R

```diff
@@ -0,0 +1,175 @@
+#' Perform a deployment "dry run"
+#'
+#' @description
+#' `r lifecycle::badge("experimental")`
+#'
+#' `dryRun()` runs your app locally, attempting to simulate what will happen
+#' when you deploy it on another machine. This isn't a 100% reliable way of
+#' discovering problems, but it offers a much faster iteration cycle, so where
+#' it does reveal a problem, it will typically make identifying and fixing it
+#' much faster.
+#'
+#' This function is still experimental, so please let us know your experiences
+#' and where we could do better:
+#' <https://github.com/rstudio/rsconnect/issues/new>
+#'
+#' ## Where it helps
+#'
+#' `dryRun()` was motivated by the two most common problems when deploying
+#' your app:
+#'
+#' * The server doesn't install all the packages your app needs to work.
+#'   `dryRun()` reveals this by using `renv::restore()` to create a project
+#'   specific library that uses only the packages that are explicitly used
+#'   by your project.
+#'
+#' * The server doesn't have environment variables you need. `dryRun()`
+#'   reveals this by removing any environment variables that you've
+#'   set in `~/.Renviron`, except for those that you declare in `envVars`.
+#'   Additionally, to help debugging it also reports whenever any env var is
+#'   used.
+#'
+#' `dryRun` will also log when you use functions that are usually best avoided
+#' in deployed code. This includes:
+#'
+#' * `rsconnect::deployApp()` because you shouldn't deploy an app from another
+#'   app. This typically indicates that you've included a file with scratch
+#'   code.
+#'
+#' * `install.packages()` and `.libPaths()` because you should rely on the
+#'   server to install and manage your packages.
+#'
+#' * `browser()`, `browseURL()`, and `rstudioapi::askForPassword()` because
+#'    they need an interactive session that your deployment server will lack.
+#'
+#' ## Current limitations
+#'
+#' * `dryRun()` currently offers no way to diagnose problems with
+#'   mismatched R/Python/Quarto/pandoc versions.
+#'
+#' * `dryRun()` doesn't help much with paths. There are two common problems
+#'   it can't help with: using an absolute path and using the wrong case.
+#'   Both of these will work locally but fail on the server. [lint()]
+#'   uses an alternative technique (static analysis) to detect many of these
+#'   cases.
+#'
```

Comment:
Another limitation is that:

* `dryRun()` doesn't help if the server does not have the system dependencies that are present 
  on the machine on which you're running this dryRun().